ddl/job_table.go

```diff
@@ -0,0 +1,485 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ddl
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pingcap/log"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/meta"
+	"github.com/pingcap/tidb/parser/model"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/sqlexec"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.uber.org/zap"
+)
+
+var (
+	addingDDLJobGeneral = "/tidb/ddl/add_ddl_job_general"
+	addingDDLJobReorg   = "/tidb/ddl/add_ddl_job_reorg"
+)
+
+func (d *ddl) insertRunningReorgJobMap(id int) {
+	d.runningReorgJobMapMu.Lock()
+	defer d.runningReorgJobMapMu.Unlock()
+	d.runningReorgJobMap[id] = struct{}{}
+}
+
+func (d *ddl) deleteRunningReorgJobMap(id int) {
+	d.runningReorgJobMapMu.Lock()
+	defer d.runningReorgJobMapMu.Unlock()
+	delete(d.runningReorgJobMap, id)
+}
+
+func (d *ddl) getGeneralJob(sess sessionctx.Context) (*model.Job, error) {
+	runningOrBlockedIDs := make([]string, 0, 10)
+	d.runningReorgJobMapMu.RLock()
+	for id := range d.runningReorgJobMap {
+		runningOrBlockedIDs = append(runningOrBlockedIDs, strconv.Itoa(id))
+	}
+	d.runningReorgJobMapMu.RUnlock()
+	var sql string
+	ts := time.Now()
+	if len(runningOrBlockedIDs) == 0 {
+		sql = "select job_meta from mysql.tidb_ddl_job where job_id in (select min(job_id) from mysql.tidb_ddl_job group by schema_id, table_id order by min(job_id)) and not reorg"
+	} else {
+		sql = fmt.Sprintf("select job_meta from mysql.tidb_ddl_job where job_id in (select min(job_id) from mysql.tidb_ddl_job group by schema_id, table_id order by min(job_id)) and not reorg and job_id not in (%s)", strings.Join(runningOrBlockedIDs, ", "))
+	}
+	rs, err := sess.(sqlexec.SQLExecutor).ExecuteInternal(context.Background(), sql)
+	if err != nil {
+		return nil, err
+	}
+	var rows []chunk.Row
+	if rows, err = sqlexec.DrainRecordSet(d.ctx, rs, 8); err != nil {
+		return nil, err
+	}
+	if err = rs.Close(); err != nil {
+		return nil, err
+	}
+	if len(rows) == 0 {
+		return nil, nil
+	}
+	jobs := make([]*model.Job, 0, 5)
+	for _, row := range rows {
+		jobBinary := row.GetBytes(0)
+		job := model.Job{}
+		err = job.Decode(jobBinary)
+		if err != nil {
+			return nil, err
+		}
+		log.Warn("get ddl job from table", zap.String("time", time.Since(ts).String()), zap.String("job", job.String()))
+		if job.Type == model.ActionDropSchema {
+			for {
```

Comment:
it finds a ```drop table``` operation that ```JobID``` is smaller than itself at the same ```SchemaID``` will not finish it. then it will try to pick another job to complete it.

Reply:
Let me take into account whether it has a better way to implement it.